### PR TITLE
Adding migration instruction from benchmarking v1 to v2

### PR DESCRIPTION
**Pull request by re-gius**
# Adding instruction to migrate benchmarking from v1 to v2

Even if the documentation for benchmarking v1 and v2 is clear and detailed, I feel that adding a migration guide from v1 to v2 would help doing it quicker. 

## Integration
This change only affects documentation, so it does not cause integration issues.

## Review Notes
I followed the migration procedure I applied in PR https://github.com/paritytech/polkadot-sdk/pull/6018 . I added everything from there, but I may be missing some extra steps that are needed in specific case, so in case you notice something please let me know.

### CODE REVIEW

File: substrate/frame/benchmarking/src/lib.rs [hunk 1]
```diff
@@ -311,6 +311,78 @@ pub use v1::*;
...
+/// 1. Change the import from `frame_benchmarking::v1::` to `frame_benchmarking::v2::*`, or
+///    `frame::benchmarking::prelude::*` under the umbrella crate;
+/// 2. Move the code inside the v1 `benchmarks! { ... }` block to the v2 benchmarks module `mod
+///    benchmarks { ... }` under the benchmarks macro (`#[benchmarks]` or `#[instance_benchmarks]`);
+/// 3. Turn each v1 benchmark into a function inside the v2 benchmarks module with the same name,
+///    having either with a blank return type or a return type compatible with `Result<(),
+///    BenchmarkError>`. For instance, `foo { ... }` can become `fn foo() -> Result<(),
+///    BenchmarkError>`. More in detail:
+///   3.1 Move all the v1 complexity parameters as [ParamRange](`v2::ParamRange`) arguments to the
+/// v2 function, and their setup code to the body of the function. For instance, `let y in 0 .. 10
```
Comment (kianenigma) [inline substrate/frame/benchmarking/src/lib.rs:328]: indentation here looks a messy.

File: substrate/frame/benchmarking/src/lib.rs [hunk 2]
```diff
@@ -311,6 +311,78 @@ pub use v1::*;
...
+/// => setup(y)?;` from v1 will give a `y: Linear<0, 10>` argument to the corresponding function in
+/// v2, while `setup(y)?;` will be moved to the body of the function;
+///   3.2 Move all the v1 setup code to the body of the v2 function;   
+///   3.3 Move the benchmarked code to the body of the v2 function under the appropriate macro
+/// attribute: `#[extrinsic_call]` for extrinsic pallet calls and `#[block]` for blocks of code;   
+///   3.4 Move the v1 verify code block to the body of the v2 function, after the benchmarked code.
```
Comment (kianenigma) [inline substrate/frame/benchmarking/src/lib.rs:334]:
```suggestion
///   3.4 Move the v1 verify code block to the body of the v2 function, after #[extrinsic_call] or #[block] attribute.
```

File: substrate/frame/benchmarking/src/lib.rs [hunk 3]
```diff
@@ -311,6 +311,78 @@ pub use v1::*;
...
+///   // size `l`
+///   #[benchmark]
+///   fn foo(l: Linear<1 .. 10_000>) -> Result<(), BenchmarkError> {
+///     let caller = funded_account::<T>(b"caller", 0);
+///     initialize_l(l);
+///
+///     #[extrinsic_call]
+///     _(RuntimeOrigin::Signed(caller), vec![0u8; l]);
+///     
+///     assert_last_event::<T>(Event::FooExecuted { result: Ok(()) }.into());
```
Comment (kianenigma) [inline substrate/frame/benchmarking/src/lib.rs:381]:
```suggestion
///     // Everything onwards will be treated as test. 
///     assert_last_event::<T>(Event::FooExecuted { result: Ok(()) }.into());
```

File: substrate/frame/benchmarking/src/lib.rs [hunk 4]
```diff
@@ -311,6 +311,78 @@ pub use v1::*;
...
+///     Ok(())
+///   }
+/// }
```
Comment (kianenigma) [inline substrate/frame/benchmarking/src/lib.rs:384]: Perhaps something about this macro that generates test tests? I think it remains unchanged for the most part.

**Review by kianenigma — APPROVED** [review]
Thank you very much!

File: substrate/frame/benchmarking/src/lib.rs (hunk 1 above)
Comment (re-gius) [reply substrate/frame/benchmarking/src/lib.rs]: True, fixed in https://github.com/paritytech/polkadot-sdk/pull/6093/commits/d80beaaa5412d8250db735f2c4300190d1a889c5

File: substrate/frame/benchmarking/src/lib.rs (hunk 2 above)
Comment (re-gius) [reply substrate/frame/benchmarking/src/lib.rs]: changed in https://github.com/paritytech/polkadot-sdk/pull/6093/commits/d80beaaa5412d8250db735f2c4300190d1a889c5

File: substrate/frame/benchmarking/src/lib.rs (hunk 3 above)
Comment (re-gius) [reply substrate/frame/benchmarking/src/lib.rs:384]: changed in https://github.com/paritytech/polkadot-sdk/pull/6093/commits/d80beaaa5412d8250db735f2c4300190d1a889c5

**Review by rockbmb — APPROVED** [review]
Great summary!

There's a construct from `v1` that doesn't exist in `v2` whose refactoring would be nice to describe here, but I don't know how it can be done in the general case.

In `v1` of the benchmarking pallet, there exists a macro named `selected_benchmark`, which is used to create a [`SelectedBenchmark`](https://github.com/paritytech/polkadot-sdk/blob/master/substrate/frame/benchmarking/src/v1.rs#L944) enum, which will contain all of the benchmarks defined inside `benchmark!` as variants.
From this `enum` we can access a specific benchmark to e.g. run it as a unit test.
Here are examples: https://github.com/paritytech/polkadot-sdk/blob/master/substrate/frame/benchmarking/src/tests.rs#L256

In #6018 there was no code using it, but in #6025, there was [one](https://github.com/paritytech/polkadot-sdk/pull/6025/files#diff-7a8630c674cd747dcd8f6a5440da01b2821c872ed2485c2b0a31906ea4e0ad56L1069) occurrence - which luckily could just be removed.
Doing a code search on `SelectedBenchmark` over the entire `polkadot-sdk` shows it is used in:
* `frame-benchmarking` itself (`src/tests.rs`, `src/v1.rs`)
* `frame-support`
* `frame-benchmarking-cli` 

I doubt moving the above pallets/crates to `v2` is a high-priority task, so describing how to migrate `SelectedBenchmark` can be done at a later time.

**Review by rockbmb — APPROVED** [review]

File: substrate/frame/benchmarking/src/lib.rs [hunk 5]
```diff
@@ -311,6 +311,83 @@ pub use v1::*;
...
+///       .. 10 => setup(y)?;` from v1 will give a `y: Linear<0, 10>` argument to the corresponding
+///       function in v2, while `setup(y)?;` will be moved to the body of the function;
+///    2. Move all the v1 setup code to the body of the v2 function;
+///    3. Move the benchmarked code to the body of the v2 function under the appropriate macro
+///       attribute: `#[extrinsic_call]` for extrinsic pallet calls and `#[block]` for blocks of
+///       code;
+///    4. Move the v1 verify code block to the body of the v2 function, after the #[extrinsic_call]
+///       or #[block] attribute.
+///    5. If the function is supposed to return a `Result<(), BenchmarkError>`, end with an `Ok(())`
+///       expression.
```
Comment (seadanda) [inline substrate/frame/benchmarking/src/lib.rs:338]:
```suggestion
///    5. If the function returns a `Result<(), BenchmarkError>`, end with `Ok(())`.
```

File: substrate/frame/benchmarking/src/lib.rs [hunk 6]
```diff
@@ -311,6 +311,83 @@ pub use v1::*;
...
 /// ```
+///
+/// ## Migrate from v1 to v2
+///
+/// If you want to migrate your code from benchmarking v1 to benchmarking v2, you may follow these
+/// steps:
+/// 1. Change the import from `frame_benchmarking::v1::` to `frame_benchmarking::v2::*`, or
+///    `frame::benchmarking::prelude::*` under the umbrella crate;
+/// 2. Move the code inside the v1 `benchmarks! { ... }` block to the v2 benchmarks module `mod
+///    benchmarks { ... }` under the benchmarks macro (`#[benchmarks]` or `#[instance_benchmarks]`);
```
[the remark this answers is not in the document]
Comment (re-gius) [reply substrate/frame/benchmarking/src/lib.rs]: added in https://github.com/paritytech/polkadot-sdk/pull/6093/commits/509a173d9bd2bdc52c7487b8ec83db25edfc6784

**Review by ggwpez — APPROVED** [review]

**Pull request merged**